src/core/execution/FakeHumanExecution.ts

```diff
@@ -437,10 +438,30 @@ export class FakeHumanExecution implements Execution {
       this.maybeSpawnStructure(UnitType.Port, 1) ||
       this.maybeSpawnStructure(UnitType.City, 2) ||
       this.maybeSpawnWarship() ||
-      this.maybeSpawnStructure(UnitType.MissileSilo, 1)
+      this.maybeSpawnStructure(UnitType.MissileSilo, 1) ||
+      this.maybeSpawnTrainStation()
     );
   }
 
+  private maybeSpawnTrainStation(): boolean {
+    if (this.player === null) throw new Error("not initialized");
+    const cities = this.player.units(
+      UnitType.City,
+      UnitType.Port,
+      UnitType.Factory,
+    );
+    const citiesWithoutStations = cities.filter(
+      (city: Unit) => !city.hasTrainStation(),
+    );
```

Comment:
You're filtering twice here, which isn't ideal because each filter allocates a new array.
```suggestion
    const citiesWithoutStations = this.player.units().filter((unit) => {
      if (unit.hasTrainStation()) return false;
      const type = unit.type();
      return (
        type === UnitType.City ||
        type === UnitType.Port ||
        type === UnitType.Factory);
    );
```